src/toolkits/coreml_export/neural_net_models_exporter.cpp

```diff
@@ -262,7 +238,136 @@ std::shared_ptr<MLModelWrapper> export_object_detector_model(
+  set_feature_optional(iou_threshold_desc);
+  model_desc->add_input()->CopyFrom(*iou_threshold_desc);
+
+  // Write FeatureDescription for the Confidence Threshold input.
+  FeatureDescription* confidence_threshold_desc = pipeline_desc->add_input();
+  set_array_feature(confidence_threshold_desc, "confidenceThreshold",
+                    confidence_threshold_description(confidence_threshold),
+                    {1});
+  set_feature_optional(confidence_threshold_desc);
+  model_desc->add_input()->CopyFrom(*confidence_threshold_desc);
 
+  // Write FeatureDescription for the Confidence output.
+  FeatureDescription* confidence_desc = pipeline_desc->add_output();
+  set_predictions_feature(confidence_desc, "confidence", max_boxes, num_classes,
+                          false, true, CONFIDENCE_STR);
+  model_desc->add_output()->CopyFrom(*confidence_desc);
+
+  // Write FeatureDescription for the Coordinates output.
+  FeatureDescription* coordinates_desc = pipeline_desc->add_output();
+  set_predictions_feature(coordinates_desc, "coordinates", max_boxes, 4, false,
+                          true, COORDINATES_STR);
+  model_desc->add_output()->CopyFrom(*coordinates_desc);
+
+  // Write FeatureDescription for the numberOfBoxes output.
+  FeatureDescription* number_of_boxes_desc = pipeline_desc->add_output();
+  set_array_feature(number_of_boxes_desc, "numberOfBoxes",
+                    "The number of valid output bounding boxes", {1});
+  model_desc->add_output()->CopyFrom(*number_of_boxes_desc);
+
+  // Write FeatureDescription for the indicesOfBoxes output.
+  FeatureDescription* indices_of_boxes_desc = pipeline_desc->add_output();
+  set_array_feature(indices_of_boxes_desc, "indicesOfBoxes",
+                    "For each output bounding box, the index of the "
+                    "corresponding input bounding box",
+                    {max_boxes});
+  model_desc->add_output()->CopyFrom(*indices_of_boxes_desc);
+}
+
+}  // namespace
+
+std::shared_ptr<MLModelWrapper> export_object_detector_model(
+    neural_net::pipeline_spec raw_pipeline, size_t num_classes,
+    size_t num_predictions, flex_list class_labels, float confidence_threshold,
+    float iou_threshold, bool include_non_maximum_suppression,
+    bool use_nms_layer) {
+  // Set up Pipeline
+  CoreML::Specification::Model model_pipeline;
+  model_pipeline.set_specificationversion(
+      CoreML::MLMODEL_SPECIFICATION_VERSION);
+  ModelDescription* pipeline_desc = model_pipeline.mutable_description();
+
+  // Adopt the model pipeline passed to us as input.
+  std::unique_ptr<CoreML::Specification::Pipeline> raw_pipeline_spec =
+      std::move(raw_pipeline).move_coreml_spec();
+  model_pipeline.mutable_pipeline()->Swap(raw_pipeline_spec.get());
+
+  if (!include_non_maximum_suppression) {
+    // Only support this case for models supporting spec version 1, which means
+    // no pipeline models.
+    ASSERT_EQ(model_pipeline.pipeline().models_size(), 1);
```

Comment:
if my reading is correct, places where we use `assert` (from `cassert`) don't throw in release builds but `ASSERT_EQ` and friends always do. i've noticed in other PRs folks have switched to `cassert`. maybe better for an offline conversation, but curious what the use conventions are here and why you chose this here and `assert` on line 358.

Reply:
Good question. These should be `ASSERT_EQ` and `ASSERT_GT`. I've been using the standard `assert` in the `neural_net` directory since I am pushing that code to be portable / standalone. But this code is in `toolkits`, which I think is more firmly dependent on the broader Turi Create code base.

Reply:
what about debug vs. release?

Reply:
What I have gathered is that the ASSERT macros are intended to catch programming errors, while we use `log_and_throw` to surface user errors. I honestly don't have context on why we have both debug-only and debug-and-release ASSERT macros. Most of this codebase seems to err on the side of compiling out only the assertions in performance-sensitive areas of the code.